webapp/application.py

```diff
@@ -121,8 +134,153 @@ def application_page(token):
+    name, candidate_id, application_id = tuple(decrypted.split("-"))
+    candidate_id, application_id = int(candidate_id), int(application_id)
+    application = harvest.get_application(application_id)
+    candidate = harvest.get_candidate(application["candidate_id"])
+    job = harvest.get_job(application["jobs"][0]["id"])
+
+    for recruiter in job["hiring_team"]["recruiters"]:
+        if recruiter["responsible"]:
+            hiring_lead = harvest.get_user(recruiter["id"])
+            break
+
+    if (
+        "candidate_id" not in application
+        or application["candidate_id"] != candidate_id
+        or not candidate["first_name"].lower() == name
+    ):
+        flask.abort(404)
+
+    candidate_name = candidate["first_name"]
+    position = application["jobs"][0]["name"]
+    hiring_lead = hiring_lead["name"]
+
+    # Sanitize and parse user input
+    email = parseaddr(flask.request.form["email"])[1]
+    candidate_email = parseaddr(candidate["email_addresses"][0]["value"])[1]
+
+    if flask.request.form["withdrawal-reason-other"]:
+        withdrawal_reason = flask.request.form["withdrawal-reason-other"]
+    else:
+        withdrawal_reason = flask.request.form["withdrawal-reason"]
+
+    # Reject if user typed the wrong email
+    if candidate_email != email:
+        return redirect(flask.request.referrer + "#wrong-email")
+
+    send_mail(
+        to_email=[candidate_email],
+        subject="Withdraw Application Confirmation",
+        message=flask.render_template(
+            "applications/_activate-email.html",
+            applicant_name=candidate_name,
+            position=position,
+            hiring_lead=hiring_lead,
+            verification_link=confirmation_token(
+                candidate_email,
+                withdrawal_reason,
+                candidate_id,
+                application_id,
+                hiring_lead,
+            ),
+        ),
+    )
+    return redirect(flask.request.referrer + "#withdrawal-requested")
+
+
+def confirmation_token(
+    email, withdrawal_reason, candidate_id, application_id, hiring_lead
+):
+    payload = {
+        "email": email,
+        "withdrawal_reason": withdrawal_reason,
+        "candidate_id": candidate_id,
+        "application_id": application_id,
+        "hiring_lead": hiring_lead,
+    }
+    token = json.dumps(payload)
+    return cipher.encrypt(token)
+
+
+# Send email if configured
+smtp_server = os.getenv("SMTP_SERVER")
+smtp_user = os.getenv("SMTP_USER")
+smtp_pass = os.getenv("SMTP_PASS")
+smtp_sender_address = os.getenv("SMTP_SENDER_ADDRESS")
+
+
+def send_mail(
+    to_email,
+    subject,
+    message,
+):
+    # import smtplib
+    msg = EmailMessage()
+    msg["Subject"] = subject
+    msg["From"] = smtp_sender_address
+    msg["To"] = ", ".join(to_email)
+    msg.set_content(message, subtype="html")
+
+    server = SMTP(smtp_server)
+    # check if smtp_server exists as it would not exist on development
+    if smtp_server:
```

Comment:
So what you're saying here is that - in local development we don't expect the SMTP server to be configured, so we won't send an actual email.

I'm worried that this will easily lead to us completely failing to realise that our SMTP server is broken and emails potentially aren't being sent.

Could we do this through a different mechanism? I'm not exactly sure what's best. Probably, this `send_mail` function should error as normal, but then at line 238 we should catch the error *only if* we're in debug mode. We could even display the email on the resulting page or something for easier debugging?

Or we could just let it error and expect anyone who really wants to test this out properly just to configure the SMTP server. Is that super unreasonable?

Reply:
yep @jpmartinspt and I tried to make it work the other day but it's blocked to send an actual email through SMTP server in local development for some reason... what you think @jpmartinspt?

Reply:
I'm almost sure we are not able to connect to the SMTP server from our machines, so I suggested we just print to message to the console when in development. I would not even try to connect when in development as it will likely just stall for as long as the connection timeout is set for. We can always QA the e-mail is actually being sent by using the demo(and maybe an integration test as part of the PR workflow) and let eventual staging/production errors with the sending land in Sentry.

Reply:
Ok. But let's:

a) do it explicitly based on debug mode (or something similar) rather than the existence of a variable
b) put a message in the actual success page to say that the email wasn't actually sent